Composer/packages/lib/indexers/src/dialogUtils/dialogChecker.ts

```diff
@@ -11,30 +11,94 @@ import { CheckerFunc } from './types';
+
+  return `${list[0]}${focused}#${type}#${list[1]}`;
+};
+
+export const checkExpression = (exp: string, required: boolean, path: string, type: string): Diagnostic | null => {
+  let message = '';
+  if (!exp && required) {
+    message = formatMessage(`is missing or empty`);
+  } else {
+    try {
+      ExpressionParser.parse(exp);
+    } catch (error) {
+      message = `${formatMessage('must be an expression:')} ${error})`;
     }
-    if (message) {
-      const diagnostic = new Diagnostic(message, '');
-      diagnostic.path = `${path}#${value.$type}#${property}`;
-      result.push(diagnostic);
+  }
+  if (message) {
+    const diagnostic = new Diagnostic(message, '');
+    diagnostic.path = createPath(path, type);
+    return diagnostic;
+  }
+
+  return null;
+};
+
+function findAllRequiredType(schema: any): { [key: string]: boolean } {
+  if (!schema) return {};
+  const types = schema.anyOf?.filter(x => x.title === 'Type');
+  const required = {};
+  if (types && types.length) {
+    types[0].required.forEach((element: string) => {
+      required[element] = true;
+    });
+  }
+
+  if (schema.required) {
+    schema.required.forEach((element: string) => {
+      required[element] = true;
+    });
+  }
+  return required;
+}
+
+export const IsExpression: CheckerFunc = (path, value, type, schema) => {
+  if (!schema) return [];
+  const diagnostics: Diagnostic[] = [];
+  const requiredTypes = findAllRequiredType(schema);
+  Object.keys(value).forEach(key => {
+    const property = value[key];
+    if (Array.isArray(property)) {
+      const itemsSchema = schema.properties[key].items;
+      if (itemsSchema?.$role === 'expression') {
+        property.forEach((child, index) => {
+          const diagnostic = checkExpression(child, !!requiredTypes[key], `${path}.${key}[${index}]`, type);
+          if (diagnostic) diagnostics.push(diagnostic);
+        });
+      } else if (itemsSchema?.type === 'object') {
+        property.forEach((child, index) => {
+          const result = IsExpression(`${path}.${key}[${index}]`, child, type, itemsSchema);
+          if (result) diagnostics.splice(0, 0, ...result);
+        });
+      }
+    } else if (get(schema.properties[key], '$role') === 'expression') {
+      const diagnostic = checkExpression(property, !!requiredTypes[key], `${path}.${key}`, type);
+      if (diagnostic) diagnostics.push(diagnostic);
     }
-    return result;
-  }, []);
+  });
+  return diagnostics;
+};
+
+//the type of 'Microsoft.ChoiceInput' has anyof schema in choices
+export const checkChoices: CheckerFunc = (path, value, type, schema) => {
+  const choices = value.choices;
+  if (typeof choices === 'string') {
+    const diagnostic = checkExpression(choices, false, `${path}.choices`, type);
+    if (diagnostic) return [diagnostic];
+  }
   return null;
 };
+
+export const checkerFuncs: { [type: string]: CheckerFunc[] } = {
+  '.': [IsExpression], //this will check all types
+  'Microsoft.ChoiceInput': [checkChoices],
```

Comment:
can we pull this valued off the SDKTypes object? `[SDKTypes.ChoiceInput]`

Reply:
Hi, @cwhitten We don't want to make indexers depends on shared because the shared has a lot of composer's util function. We want to make indexers be used outside the composer.

Maybe we can create a library called schema in the future. 
@boydc2014 do you have any comments about these?

Reply:
The schema needs to be injected in the running application, it needs to be "stateless" and exported from one place. Pulling it from `shared` is the right thing to do, and we can extract this further to come from a less heavy dependency. Making the change now will make that effort much easier.

Reply:
Pull this from one single place instead of  hard coding it is definitely a good practice, and the concern of "this shared package is too big" is also valid. I agree @cwhitten that we should  pull it now, and seek to split this "shared" into less heavy modules later.  The name `shared` already indicate it's not something we thought too much about :)